src/Database/LSMTree/Internal/BloomFilter.hs

```diff
@@ -125,3 +129,57 @@ hGetByteArrayExactly hfs h len = do
     _   <- hGetBufExactly hfs h buf 0 (fromIntegral len)
     P.unsafeFreezeByteArray buf
 
+-- Bulk insert
+-----------------------------------------------------------
+
+{-# INLINE bloomInserts #-}
+bloomInserts ::
+     forall a s.
+     BF.Hashable a
+  => BF.MBloom s a
+  -> (Int -> ST s a)
+  -> Int
+  -> ST s ()
+bloomInserts bloom key n =
+    P.newPrimArray 0x10 >>= body
+  where
+    body :: P.MutablePrimArray s (BF.Hashes a) -> ST s ()
+    body !buf = prepareProbes 0 0
+      where
+        prepareProbes :: Int -> Int -> ST s ()
+        prepareProbes !i !i_w
+          | i_w < 0x0f && i < n = do
+              k <- key i
+              let !kh = BF.hashes k
+              BF.prefetchInsert bloom kh
+              P.writePrimArray buf i_w kh
+              prepareProbes (i+1) (i_w+1)
+
+          | n > 0     = insertProbe 0 0 i_w
+          | otherwise = return ()
+
+        insertProbe :: Int -> Int -> Int -> ST s ()
+        insertProbe !i !i_r !i_w = do
+            kh <- P.readPrimArray buf i_r
+            BF.insertHashes bloom kh
+            nextProbe i i_r i_w
+
+        nextProbe :: Int -> Int -> Int -> ST s ()
+        nextProbe !i !i_r !i_w
+          | i < n = do
+              k <- key i
+              let !kh = BF.hashes k
+              BF.prefetchInsert bloom kh
+              P.writePrimArray buf i_w kh
+              insertProbe
+                (i+1)
+                ((i_r + 1) .&. 0x0f)
+                ((i_w + 1) .&. 0x0f)
+
+          | ((i_r + 1) .&. 0x0f) /= i_w =
+              insertProbe
+                i
+                ((i_r + 1) .&. 0x0f)
+                i_w
```

Comment:
IIUC, this case kicks in when there are no more keys to prefetch, so we just have to do all the remaining probe inserts until we hit the end of the rolling buffer (`i_r == i_w`)

Reply:
Yes exactly. I should probably add a comment to that effect.

Reply:
Documented.